pkg/kubelet/metrics/metrics.go

```diff
@@ -30,30 +35,77 @@ var (
 			Help:      "Image pull latency in microseconds.",
 		},
 	)
-	// TODO(vmarmol): Implement.
-	// TODO(vmarmol): Split by source?
-	PodCount = prometheus.NewGauge(
-		prometheus.GaugeOpts{
-			Subsystem: kubeletSubsystem,
-			Name:      "pod_count",
-			Help:      "Number of pods currently running.",
-		},
-	)
-	// TODO(vmarmol): Implement.
-	// TODO(vmarmol): Split by source?
-	ContainerCount = prometheus.NewGauge(
-		prometheus.GaugeOpts{
-			Subsystem: kubeletSubsystem,
-			Name:      "container_count",
-			Help:      "Number of containers currently running.",
-		},
-	)
 	// TODO(vmarmol): Containers per pod
 	// TODO(vmarmol): Latency of pod startup
 	// TODO(vmarmol): Latency of SyncPods
 )
 
-func init() {
+var registerMetrics sync.Once
+
+// Register all metrics.
+func Register(containerCache dockertools.DockerCache) {
 	// Register the metrics.
-	prometheus.MustRegister(ImagePullLatency)
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(ImagePullLatency)
+		prometheus.MustRegister(newPodAndContainerCollector(containerCache))
+	})
+}
+
+func newPodAndContainerCollector(containerCache dockertools.DockerCache) *podAndContainerCollector {
+	return &podAndContainerCollector{
+		containerCache: containerCache,
+	}
+}
+
+// Custom collector for current pod and container counts.
+type podAndContainerCollector struct {
+	// Cache for accessing information about running containers.
+	containerCache dockertools.DockerCache
+}
+
+// TODO(vmarmol): Split by source?
+var (
+	runningPodCountDesc = prometheus.NewDesc(
+		prometheus.BuildFQName("", kubeletSubsystem, "running_pod_count"),
+		"Number of pods currently running",
+		nil, nil)
+	runningContainerCountDesc = prometheus.NewDesc(
+		prometheus.BuildFQName("", kubeletSubsystem, "running_container_count"),
+		"Number of containers currently running",
+		nil, nil)
+)
+
+func (self *podAndContainerCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- runningPodCountDesc
+	ch <- runningContainerCountDesc
+}
+
+func (self *podAndContainerCollector) Collect(ch chan<- prometheus.Metric) {
+	runningContainers, err := self.containerCache.RunningContainers()
```

Comment:
Ok. Is it not possible to have a wrapper around this in kubelet core that provides pods instead of containers. As of now, I see quite a bit of duplicated code in kubelet around docker containers handling. This is not required for this PR, but having one API for getting pods might help in the long run.